plugin/pkg/scheduler/scheduler.go

```diff
@@ -230,6 +239,97 @@ func (sched *Scheduler) preempt(preemptor *v1.Pod, scheduleErr error) (string, e
+		// TODO: can we distingush between fully bound and waiting for binding to complete?
+		// If it's waiting for binding to complete, then we can return early here instead of
+		// spawning a go routine for bindVolumes to do nothing...
+		bindingRequired, err := sched.config.VolumeBinder.Binder.AssumePodVolumes(assumed, host)
+		if err != nil {
+			sched.config.Error(assumed, err)
+			sched.config.Recorder.Eventf(assumed, v1.EventTypeWarning, "FailedScheduling", "AssumePod failed: %v", err)
+			sched.config.PodConditionUpdater.Update(assumed, &v1.PodCondition{
+				Type:    v1.PodScheduled,
+				Status:  v1.ConditionFalse,
+				Reason:  "SchedulerError",
+				Message: err.Error(),
+			})
+			return err
+		}
+		if bindingRequired {
+			if sched.config.Ecache != nil {
+				invalidPredicates := sets.NewString("NoVolumeBindConflict")
+				sched.config.Ecache.InvalidateCachedPredicateItemOfAllNodes(invalidPredicates)
+			}
+
+			sched.config.VolumeBinder.BindQueue.Add(assumed)
+
+			// bindVolumesWorker() will update the Pod object to put it back in the scheduler queue
+			return fmt.Errorf("Volume binding started, waiting for completion")
+		}
+	}
+	return nil
+}
+
+// bindVolumesWorker() processes pods queued in assumeAndBindVolumes() and tries to
+// make the API update for volume binding.
+// This function runs forever until the volume BindQueue is closed.
+func (sched *Scheduler) bindVolumesWorker() {
+	workFunc := func() bool {
+		keyObj, quit := sched.config.VolumeBinder.BindQueue.Get()
+		if quit {
+			return true
+		}
+		defer sched.config.VolumeBinder.BindQueue.Done(keyObj)
+
+		assumed, ok := keyObj.(*v1.Pod)
+		if !ok {
+			glog.V(4).Infof("Object is not a *v1.Pod")
+			return false
+		}
+
+		// TODO: add metrics
+		var reason string
+
+		glog.V(5).Infof("Trying to bind volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name)
+
+		// The Pod is always sent back to the scheduler afterwards.
+		err := sched.config.VolumeBinder.Binder.BindPodVolumes(assumed)
+		if err != nil {
+			glog.V(1).Infof("Failed to bind volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name, err)
+			reason = "VolumeBindingFailed"
+		} else {
+			glog.V(4).Infof("Successfully bound volumes for pod \"%v/%v\"", assumed.Namespace, assumed.Name)
+			reason = "VolumeBindingWaiting"
+			err = fmt.Errorf("Volume binding started, waiting for completion")
+		}
+
+		// Always fail scheduling regardless of binding success.
+		// The Pod needs to be sent back through the scheduler to:
+		// * Run through other volume predicates that require volumes to be bound and have not
+		//   been integrated into the new volume binder.
+		// * Retry volume binding if it fails.
+		// * Retry volume binding if dynamic provisioning fails.
+		// * Bind the Pod to the Node once all volumes are bound.
+		sched.config.Error(assumed, err)
+		sched.config.Recorder.Eventf(assumed, v1.EventTypeWarning, "FailedScheduling", "%v", err)
```

Comment:
Is it a warning for the waiting case?

Reply:
I guess if the reason is "VolumeBindingFailed", then it should be a warning.  Otherwise, it should be normal.  But it's also not a "successful" event.